<h1 style="text-align: center">
<div style="color: #DD3403; font-size: 60%">Data Science DISCOVERY MicroProject</div>
<span style="">MicroProject: Building a Scene Recognition Model form Video Frames</span>
<div style="font-size: 60%;"><a href="https://discovery.cs.illinois.edu/microproject/video-frame-scene-recognition-model/">https://discovery.cs.illinois.edu/microproject/video-frame-scene-recognition-model/</a></div>
</h1>

<hr style="color: #DD3403;">

## Data Source: Frames of a Video

Visual images are an important part of all media and Data Scientists are often using images as data sources.  In this MicroProject, you will create a simple model to detect the amount of time spent in two different "scenes" we used when creating office-hour style videos for Data Science DISCOVERY.

*This MicroProject was inspired by a podcast that we recently recorded with the team from the Center for Innovation in Teaching and Learning who helped produce our video.  To learn the background and hear from Karle and Wade about the journey of creating DISCOVERY, go over and listen to our episode on the "Teach Talk Listen Learn Podcast" where talk with TTLL host Bob Dignan and our CITL video producer Eric Schumacher: https://citl.illinois.edu/citl-101/teaching-learning/teach-talk-listen-learn*


### Loading a Video Frame

We have provided you with one frame every second from our video [*"Outliers Impact on Correlation (m6-02b)"*](https://www.youtube.com/watch?v=bd6hQ2UcIJc) that is used as part of our [DISCOVERY lecture covering Correlation](https://discovery.cs.illinois.edu/learn/Towards-Machine-Learning/Correlation/).

The `skimage` library is commonly used to load image data into Python.  Specifically:

- `skimage.io.imread(filename)` will read a filename and return the pixel color for every pixel in the image.
- To use the `imread` function, you will need to either do one of the following:

    1. Either import all of `sklearn` by using the import line `import sklearn`.  After importing all of `sklearn`, you will call the function using it's fully qualified name: `skimage.io.imread(filename)`.
    
    **OR**
    
    2. Import only the `imread` function by using the import line `from sklearn.io import imread`.  After importing only `imread`, you will call the function directly: `imread(filename)`

#### Read Pixel Data for `frames/frame_0001.jpg`

We have provided a `frames` directory with all of the frames.  In the following cell, store the pixel color data from the file named `frames/frame_0001.png` image in the variable `pixels` by using the `imread` function:


In [42]:
from skimage.io import imread
pixels = imread("frames/frame_0001.jpg")

### 🔬 Checkpoint Tests 🔬

In [43]:
### TEST CASE for Reading Video Frames
tada = "\N{PARTY POPPER}"

assert("pixels" in vars())
assert(pixels.shape == (360, 640, 3))
assert(pixels[0][0][0] == 91)

print(f"{tada} All Tests Passed! {tada}")

🎉 All Tests Passed! 🎉


<hr style="color: #DD3403;">

## Part 1: Storing Average Pixel Color

The **shape** of your data is the `rows` by `columns` by `color values` as 3-dimensional list.  Here's a formatted view of your `pixels` data:

```
[
  [ [91, 83, 80], [91, 83, 80], [91, 83, 80] ], ... ],   # Row #1
  [ [91, 83, 80], [91, 83, 80], [91, 83, 80] ], ... ],   # Row #2
  ...                                                    # ...
]
```

The current shape of `pixels` is 360 rows by 640 columns by 3 colors.  Each of the three colors represent the three color channels on a screen: red, green, and blue.

Using `pixel.mean()`, we find the average color grouping **ALL** the color channels (combining blues and reds and greens together).  Try it out:


In [44]:
pixels.mean()

72.18011863425926

In [45]:
pixels = pixels.reshape(-1, 3)
pixels


array([[ 91,  83,  80],
       [ 91,  83,  80],
       [ 91,  83,  80],
       ...,
       [162, 131, 110],
       [162, 131, 110],
       [162, 131, 110]], dtype=uint8)

To find the average of each color channel, the `pixels.resize(-1, 3).mean(axis=0)` function will find the mean of everything **except** the color channels.  Check out the new mean:

In [46]:
pixels.mean(axis=0)

array([88.65917535, 67.45620226, 60.4249783 ])

### Puzzle 1.1: Finding the Average Color of One Image

Store `pixel`'s average red value in `r`, average green value in `g`, and average blue value in `b`:

In [47]:
r, g, b = pixels.mean(axis=0)

In [48]:
### TEST CASE for Puzzle 1.1: Finding the Average Color of One Image
tada = "\N{PARTY POPPER}"

import math
assert("r" in vars())
assert("g" in vars())
assert("b" in vars())
assert(math.isclose(r, 88.65917534722222))
assert(math.isclose(g, 67.45620225694445))
assert(math.isclose(b, 60.42497829861111))

print(f"{tada} All Tests Passed! {tada}")

🎉 All Tests Passed! 🎉


### Puzzle 1.2: Finding the Average Color of All Images

The following code loops through every file in the `frames` directory -- this will include `frame_0001.jpg` (like you analyzed already) and also `frame_0002.jpg`, `frame_0003.jpg`, and all 300+ frames!

Create a DataFrame where each row is one frame with the following four columns:
- `frame`, the filename of the frame
- `r`, the average red color of the frame
- `g`, the average green color of the frame
- `b`, the average blue color of the frame

The structure of the code should be nearly identical to writing a simulation.  Instead of creating random variables for your real world data, your real world data will be the filename, and the average color values.

- See: https://discovery.cs.illinois.edu/learn/Simulation-and-Distributions/Simple-Simulations-in-Python/

In [49]:
import glob
import os
import pandas as pd

data = []
for frame in glob.glob(os.path.join("frames", "*.jpg")): 
  # `frame`` contains the filename of the frame (ex: "frames/frame_0001.jpg").  Use it for `imread` to read the frame image data.
  pixels = imread(frame)
  pixels = pixels.reshape(-1,3)
  r,g,b=pixels.mean(axis=0)
  frame_name = os.path.basename(frame)
  data.append({'frame':frame_name, 'r':r, 'g':g, 'b':b})
df = pd.DataFrame(data)
df

frame          r          g          b
0    frame_0001.jpg  88.659175  67.456202  60.424978
1    frame_0002.jpg  88.697865  67.453529  60.475660
2    frame_0003.jpg  88.028351  66.913845  60.064592
3    frame_0004.jpg  88.825629  67.340347  60.491645
4    frame_0005.jpg  88.211714  66.979661  59.983173
..              ...        ...        ...        ...
325  frame_0326.jpg   7.470391   7.473355   7.479188
326  frame_0327.jpg   7.469779   7.472743   7.478576
327  frame_0328.jpg   7.480234   7.481519   7.487826
328  frame_0329.jpg   7.480004   7.481289   7.487595
329  frame_0330.jpg   4.657478   4.658776   4.665082

[330 rows x 4 columns]

### 🔬 Checkpoint Tests 🔬

In [50]:
### TEST CASE for Puzzle 1.2: Finding the Average Color of All Images
tada = "\N{PARTY POPPER}"

import math
assert("df" in vars())
assert(len(df) == 330)
assert("r" in df)
assert("g" in df)
assert("b" in df)
assert("frame" in df)
assert( abs( df[ df.frame.str.endswith("_0001.jpg") ]["r"].sum() - 88 ) < 1 )

print(f"{tada} All Tests Passed! {tada}")

🎉 All Tests Passed! 🎉


<hr style="color: #DD3403;">

## Part 2: Create a Simple Classifier

In the DISCOVERY lecture videos, there are two primary "scenes" in the video:

1. **"Office Hours Studio Scene"**, where Karle and Wade are talking to each other and the audience,

2. **"Notebook Scene"**, where the notebook is displayed

View the `frames` folder on your computer and find **at least three more frames** that are in the "office hours studio scene" and **at least three more frames** that are in the "notebook scene".  Add the frames you found to the list below:

In [51]:
# List of at least four office hour frames by the filename's frame number:
office_hour_frames = [1,2,3,4]

# List of at least four notebook frames by the filename's frame number:
notebook_frames = [30,31,32,34]

### Observing the Average Colors of Your Frames

The following code uses your sample frames to display the average color values for your selected frames:

In [52]:
print("== Office Hour Frames ==")
print( df[ df["frame"].isin( [f"frames\\frame_{frame:04d}.jpg" for frame in office_hour_frames] ) ] )
print()
print("== Notebook Frames ==")
print( df[ df["frame"].isin( [f"frames\\frame_{frame:04d}.jpg" for frame in notebook_frames] ) ] )

== Office Hour Frames ==
Empty DataFrame
Columns: [frame, r, g, b]
Index: []

== Notebook Frames ==
Empty DataFrame
Columns: [frame, r, g, b]
Index: []


### Create Your Classifier Function

A **classifier function** is a function that takes data and gives a classification for that data.  Create a new function, `classifyFrame` that receives an `r`, `g`, and `b` value.

Using information from your frames above, have the function return the string `"office hour"` or `"notebook"` based on the values of `r`, `g`, and `b`.

**IMPORTANT**: Make sure your classifier can handle **ANY** input -- even frames you have not seen before!  For example, you might decide that you will call a frame an `"office hour"` frame if the sum of `r`, `g` and `b` is greater than 100 and otherwise it's a `"notebook"` scene.

In [64]:
def classifyFrame(r, g, b):
    sum = r + g + b
    if (sum > 600):
        return "notebook"
    else:
        return "office hour"

### 🔬 Checkpoint Tests 🔬

In [65]:
### TEST CASE for Part 2: Create a Simple Classifier
tada = "\N{PARTY POPPER}"

r = classifyFrame(0, 0, 0)
assert(r == "notebook" or r == "office hour")

r = classifyFrame(255, 255, 255)
assert(r == "notebook" or r == "office hour")

r = classifyFrame(0, 255, 255)
assert(r == "notebook" or r == "office hour")

r = classifyFrame(255, 255, 0)
assert(r == "notebook" or r == "office hour")

print(f"{tada} All Tests Passed! {tada}")

🎉 All Tests Passed! 🎉


<hr style="color: #DD3403;">

## Part 3: Using Your Classifier!

Now that we have a classifier, we should run it on every frame!

The following cell runs your `classifyFrame` classifier on every frame and adds a new column `scene` and displayed 20 random rows:

In [66]:
df["scene"] = df.apply(lambda row: classifyFrame(row.r, row.g, row.b), axis=1)
df.sample(20)

frame           r           g           b        scene
14   frame_0015.jpg  106.867882   53.494240   45.379852  office hour
71   frame_0072.jpg  230.912070  230.171858  230.673008     notebook
257  frame_0258.jpg  243.963160  243.130226  241.545473     notebook
161  frame_0162.jpg  229.838398  228.958073  226.643390     notebook
244  frame_0245.jpg  244.158095  243.408286  241.718056     notebook
285  frame_0286.jpg  243.152157  242.094431  240.853064     notebook
204  frame_0205.jpg  238.592704  237.855078  235.574896     notebook
7    frame_0008.jpg  180.616185   52.125681   37.592253  office hour
173  frame_0174.jpg   89.999431   70.290955   61.936580  office hour
54   frame_0055.jpg  233.135265  232.561944  232.752548     notebook
33   frame_0034.jpg  237.047144  236.413112  236.723785     notebook
103  frame_0104.jpg  230.872882  230.199062  227.862708     notebook
316  frame_0317.jpg  233.539870   66.995486   47.766740  office hour
9    frame_0010.jpg  108.020807   53.174227   44.903989  office hour
131  frame_0132.jpg   87.254870   67.507188   60.424058  office hour
74   frame_0075.jpg  230.714566  229.925187  230.472669     notebook
216  frame_0217.jpg   87.949683   66.988277   61.482010  office hour
64   frame_0065.jpg  231.244536  230.689201  230.925812     notebook
303  frame_0304.jpg   90.062201   70.626554   63.144909  office hour
280  frame_0281.jpg  243.228594  242.192652  240.914119     notebook

### 🔬 Checkpoint Tests 🔬

In [67]:
### TEST CASE for Part 3: Using Your Classifier
tada = "\N{PARTY POPPER}"

assert("scene" in df)

assert(len(df[ df.scene == "notebook" ]) > 100)
assert(len(df[ df.scene == "office hour" ]) > 75)
assert(len(df[ df.scene == "notebook" ]) + len(df[ df.scene == "office hour" ]) == len(df))

assert( len( df[ (df.frame.str.endswith("0001.jpg")) & (df.scene == "office hour") ] ) == 1 )
assert( len( df[ (df.frame.str.endswith("0306.jpg")) & (df.scene == "office hour") ] ) == 1 )
assert( len( df[ (df.frame.str.endswith("0081.jpg")) & (df.scene == "notebook") ] ) == 1 )
assert( len( df[ (df.frame.str.endswith("0191.jpg")) & (df.scene == "notebook") ] ) == 1 )

print(f"{tada} All Tests Passed! {tada}")

🎉 All Tests Passed! 🎉


## Observing Results

In the next 5 cells, we display a frame and you'll run code to check what your classifier classified the frame as being!  Make sure to run the code for each frame:

### Frame #0001: Office Hours

In [ ]:
df[ df.frame.str.endswith("0001.jpg") ]

frame          r          g          b        scene
0  frame_0001.jpg  88.659175  67.456202  60.424978  office hour

![Frame 0001](frames/frame_0001.jpg)

### Frame #0081: Notebook

In [ ]:
df[ df.frame.str.endswith("0081.jpg") ]

frame           r           g          b        scene
80  frame_0081.jpg  230.721385  229.915091  230.48303  office hour

![Frame 0001](frames/frame_0081.jpg)

### Frame #0191: Notebook

In [ ]:
df[ df.frame.str.endswith("0191.jpg") ]

frame           r           g           b        scene
190  frame_0191.jpg  233.117088  232.354644  230.103359  office hour

![Frame 0001](frames/frame_0191.jpg)

### Frame #0306: Office Hours

In [ ]:
df[ df.frame.str.endswith("0306.jpg") ]

frame          r          g         b        scene
305  frame_0306.jpg  89.403867  70.149223  62.83901  office hour

![Frame 0001](frames/frame_0306.jpg)

### Frame #0320: Data Science Duo Logo???

What did you classify the DUO logo as?  It's nether one, but we don't have that option!

In [ ]:
df[ df.frame.str.endswith("0320.jpg") ]

frame           r          g          b        scene
319  frame_0320.jpg  221.227565  71.838433  54.457305  office hour

![Frame 0001](frames/frame_0320.jpg)

### Frame #328: Video Credits

What did you classify the DUO logo as?  It's another tricky one!


In [ ]:
df[ df.frame.str.endswith("0328.jpg") ]

frame         r         g         b     scene
327  frame_0328.jpg  7.480234  7.481519  7.487826  notebook

![Frame 0328](frames/frame_0328.jpg)

<hr style="color: #DD3403;">

## Part 4: Update Your Classifier to Account with an "Other" Category

Create a second classifier -- `classifyFrame2` -- that returns either `"notebook"`, `"office hour"` or `"other"`.  Your classifier should correctly handle the "Data Science Duo" (ex: #0320) frames and the "Credit" frames (ex: #0328).

In [ ]:
def classifyFrame2(r, g, b):
  # Return either "office hour", "notebook", or "other" based on the values of `r`, `g`, and `b`.
  color = r+b+g
  if color <38:
    return "other"
  elif color >650:
    return "notebook"
  elif 200 <= color <=250:
    return "office hour"
  else:
    return "other"

## Apply your `classifyFrame2` function

Using `classifyFrame2`, this code replaces the value in the column `scene` with your `classifyFrame2` classification function.  The output of this cell shows the last frames of the video, which we expect to be `"other"`:

In [ ]:
df["scene"] = df.apply(lambda row: classifyFrame2(row.r, row.g, row.b), axis=1)
df.tail(20)

frame           r          g          b        scene
310  frame_0311.jpg   89.052721  70.084679  62.528559  office hour
311  frame_0312.jpg   89.577539  70.261745  62.894870  office hour
312  frame_0313.jpg   89.365169  70.192526  62.526467  office hour
313  frame_0314.jpg   89.240360  70.183095  62.777127  office hour
314  frame_0315.jpg   89.053277  70.115230  62.639093  office hour
315  frame_0316.jpg  227.706259  67.024722  48.608728        other
316  frame_0317.jpg  233.539870  66.995486  47.766740        other
317  frame_0318.jpg  227.335317  67.101398  48.556484        other
318  frame_0319.jpg  221.847739  72.007214  54.584670        other
319  frame_0320.jpg  221.227565  71.838433  54.457305        other
320  frame_0321.jpg  218.887899  80.898411  63.118472        other
321  frame_0322.jpg  200.769965  41.149089  18.975074        other
322  frame_0323.jpg  201.348303  42.190686  19.636580        other
323  frame_0324.jpg  201.755673  42.935651  20.674631        other
324  frame_0325.jpg    0.025247   0.030256   0.038095        other
325  frame_0326.jpg    7.470391   7.473355   7.479188        other
326  frame_0327.jpg    7.469779   7.472743   7.478576        other
327  frame_0328.jpg    7.480234   7.481519   7.487826        other
328  frame_0329.jpg    7.480004   7.481289   7.487595        other
329  frame_0330.jpg    4.657478   4.658776   4.665082        other

### 🔬 Checkpoint Tests 🔬

In [ ]:
### TEST CASE for Part 4: Update Your Classifier to Account with an Other Category
tada = "\N{PARTY POPPER}"

assert("scene" in df)

assert(len(df[ df.scene == "notebook" ]) > 100)
assert(len(df[ df.scene == "office hour" ]) > 75)
assert(len(df[ df.scene == "other" ]) >= 15)
assert(len(df[ df.scene == "other" ]) <= 18)   # Okay to classify the intro screens as well, but not any others.
assert(len(df[ df.scene == "notebook" ]) + len(df[ df.scene == "office hour" ]) + len(df[ df.scene == "other" ]) == len(df))

assert( len( df[ (df.frame.str.endswith("0001.jpg")) & (df.scene == "office hour") ] ) == 1 )
assert( len( df[ (df.frame.str.endswith("0306.jpg")) & (df.scene == "office hour") ] ) == 1 )
assert( len( df[ (df.frame.str.endswith("0081.jpg")) & (df.scene == "notebook") ] ) == 1 )
assert( len( df[ (df.frame.str.endswith("0191.jpg")) & (df.scene == "notebook") ] ) == 1 )
assert( len( df[ (df.frame.str.endswith("0317.jpg")) & (df.scene == "other") ] ) == 1 )
assert( len( df[ (df.frame.str.endswith("0325.jpg")) & (df.scene == "other") ] ) == 1 )
assert( len( df[ (df.frame.str.endswith("0328.jpg")) & (df.scene == "other") ] ) == 1 )

print(f"{tada} All Tests Passed! {tada}")

🎉 All Tests Passed! 🎉


<hr style="color: #DD3403;">

## Submission

You're almost done!  All you need to do is to commit your lab to GitHub and run the GitHub Actions Grader:

1.  ⚠️ **Make certain to save your work.** ⚠️ To do this, go to **File => Save All**

2.  After you have saved, exit this notebook and return to https://discovery.cs.illinois.edu/microproject/video-frame-scene-recognition-model/ and complete the section **"Commit and Grade Your Notebook"**.

3. If you see a 100% grade result on your GitHub Action, you've completed this MicroProject! 🎉